tests/schema/fail/invalid-components-object.yaml

```diff
@@ -0,0 +1,16 @@
+openapi: 3.2.0
+info:
+  title: API
+  version: 1.0.0
+paths: true
+components:
+  schemas: true
+  responses: true
+  parameters: true
+  examples: true
+  requestBodies: true
+  headers: true
+  securitySchemes: true
+  links: true
+  callbacks: true
+  pathItems: true
```

Comment:
Does the JSON Schema implementation actually check every single one of these?  Many if not most stop evaluating on the first error.  This is why most of the time negative test cases test only one condition at a time while postive test cases can test many conditions.

Also, why is `paths: true`?

Reply:
> Yes, it does. This single fail test case covers a lot of branches in a negative way.

How is that done? If you are just checking for validation failure of the entire document, you don't know if only one of those invalid properties generated an error, or some of them, or all of them.  Is the test suite counting failures? I don't see any way of indicating that per test.

Reply:
Our test suite runs the schema validation tool on a set of „pass“ and „fail“ instances, one by one.

The schema validation tool evaluates the schema on a given instance, reports all failed checks as defined for the [BASIC output structure](https://json-schema.org/draft/2020-12/json-schema-core#section-12.4.2), collects all touched schema branches, and then decides whether the instance passes or fails validation.

The test suite then checks whether the instance passes or fails as expected.

If all test cases work as expected, the test suite succeeds.

Finally the overall coverage of all schema branches across all evaluated instances is checked against the threshold of 100%

Reply:
"collects all touched schema branches" is that just the top level of the schema?  so if there is a deeper `properties` - for example testing something deep within a `parameters` object - that won't be seen?

It would be useful to publish this algorithm so that the test suite could be used by other implementations as well for validating their correctness.

For example, the JSON Schema Test Suite (https://github.com/json-schema-org/JSON-Schema-Test-Suite) includes documentation describing the structure of the test data and how to use it with any implementation, and I pull that into my JSON Schema implementation using git submodules which are run when extended testing is enabled.

Reply:
> It would be useful to publish this algorithm so that the test suite could be used by other implementations as well for validating their correctness.

We are only using the publicly available tool https://github.com/hyperjump-io/json-schema-coverage.

Reply:
> "collects all touched schema branches" is that just the top level of the schema?

No, the tool seems to consider all nesting levels, I had to add a lot of test cases to cover all branches.

Reply:
@karenetheridge Did I answer all your questions?

Reply:
> You followed up to a resolved conversation between Henry and me without unresolving it.

Ok, yes, sorry!

> We are only using the publicly available tool https://github.com/hyperjump-io/json-schema-coverage.

Ok, I'll pursue that with them.